census_view.cpp

```diff
@@ -1677,6 +2845,89 @@ void CensusDVCView::UponDeleteCells(wxCommandEvent&)
     status() << "Delete: " << timer.stop().elapsed_msec_str() << std::flush;
 }
 
+void CensusGridView::UponDeleteCells(wxCommandEvent&)
+{
+    auto const n_items = grid_table_->GetRowsCount();
+    auto rows = grid_window_->GetSelectedRows();
+    // If no cell is selected then use the current row
+    if (rows.empty())
+        {
+        rows.push_back(grid_window_->GetGridCursorRow());
+        }
+    auto const n_sel_items = lmi::ssize(rows);
+
+    if(n_items == n_sel_items)
+        {
+        warning()
+            << "Cannot delete all cells. A census must always"
+            << " contain at least one cell."
+            << LMI_FLUSH
+            ;
+        return;
+        }
+
+    LMI_ASSERT(n_sel_items < n_items);
+
+    std::ostringstream oss;
+    oss
+        << "Irrevocably delete "
+        << n_sel_items
+        << " of "
+        << n_items
+        << " cells?"
+        ;
+    int yes_or_no = wxMessageBox
+        (oss.str()
+        ,"Confirm deletion"
+        ,wxYES_NO | wxICON_QUESTION
+        );
+    if(wxYES != yes_or_no)
+        {
+        return;
+        }
+
+    wxBusyCursor reverie;
+    Timer timer;
+
+    auto const cursor_col = grid_window_->GetGridCursorCol();
+
+    std::sort(rows.begin(), rows.end());
+
+    LMI_ASSERT(lmi::ssize(cell_parms()) == n_items);
+
+    wxGridUpdateLocker grid_update_locker(grid_window_);
+    grid_window_->ClearSelection();
+
+    int items_left = n_sel_items;
+    while(items_left > 0)
+        {
+        int cur_item = items_left - 1;
+        for(; 0 < cur_item; --cur_item)
+            {
+            if(rows[cur_item] - 1 != rows[cur_item - 1])
+                break;
+            }
+        int const count = items_left - cur_item;
+        items_left -= count;
+
+        auto const first = cell_parms().begin() + cur_item;
+        cell_parms().erase(first, first + count);
+        grid_window_->DeleteRows(cur_item, count);
+        }
```

Comment:
I think that this is just trying to reconstruct the contiguous spans of rows, right?

If so, this should be replaced with `GetSelectedBlocks()` which already returns this directly. I.e. we should write a loop iterating over `wxGridBlockCoords` returned by iterating over it.

Reply:
`GetSelectedBlocks()` can return owerlapped blocks (i.e. rows 0-3, 1-4; which in `lmi` code will be united to a single  block 0-4). Also we should sort the blocks any way to delete from the biggest to the lowest.

Reply:
Unfortunately `GetSelectedRows()` itself didn't even behave correctly in this case, see https://github.com/wxWidgets/wxWidgets/pull/1874

Reply:
FYI, I've rewritten this code using the new functions from wxWidgets/wxWidgets#1877.